### PR TITLE
Correct IDE diagnostics and lifetimes for editing with #r package references

### PR DESCRIPTION
**Pull request by dsyme**
This extends https://github.com/dotnet/fsharp/pull/10159 with some basic improvements to correct how and when we create DependencyProvider

1. Only create DependencyProvider once per project, and one overall for scripts

2. Don't register endless chains of dynamic resolution handlers for the compiler and IDE cases

3. Gives diagnostics and failures in the IDE 

   image

4. Gives diagnostics and failures in F# compilation of scripts

   image

5. Gives diagnostics and failures in F# Interactive execuation of scripts

   image

6. Give tooltips (this was a feature lost on VS2017 or VS2019 for #r references in scripts, and easy to resurrect and augment to give information for package references).  Note how the text shows the exact package versions and DLLs caused by the reference

   image

   image

   image

Note that script editing is still always .NET Framework in this release, sadly - there are obviously many scripts where this doesn't matter but it's something to be aware of - we will be able to correct this somehow in the next release, it's too important not to.

image

### CODE REVIEW

File: src/fsharp/fsi/fsi.fs [hunk 1]
```diff
@@ -757,6 +761,9 @@ type internal FsiCommandLineOptions(fsi: FsiEvaluationSessionHostConfig, argv: s
             stopProcessingRecovery e range0; failwithf "Error creating evaluation session: %A" e
         inputFilesAcc
 
+    // We need a dependency provider with native resolution.  Managed resolution is handled by generated `#r`
+    let dependencyProvider = new DependencyProvider(null, NativeResolutionProbe(tcConfigB.GetNativeProbingRoots))
```
Comment (cartermp) [inline src/fsharp/fsi/fsi.fs:765]: What is `null`? Should this be an optional argument?

Comment (cartermp) [reply src/fsharp/fsi/fsi.fs:765]: actually this could just be a call to the overload without `null`

**Comment by cartermp** [comment]
Surface area needs update

**Comment by cartermp** [comment]
This looks good subject to CI passing.

**Comment by dsyme** [comment]
Note to self - tests needs adjusting

```
  X FSharp.Compiler.Scripting.UnitTests.InteractiveTests.Script with #i and forgot to add quotes(code: "#i \"nuget:foo\" ", error: "input.fsx (1,1)-(1,15) interactive error Invalid U"...) [514ms]
  X FSharp.Compiler.Scripting.UnitTests.InteractiveTests.Eval script with invalid PackageName should fail immediately and resolve one time only [2s 882ms]
  X FSharp.Compiler.Scripting.UnitTests.InteractiveTests.#i to a directory that doesn't exist [382ms]
  X FSharp.Compiler.Scripting.UnitTests.InteractiveTests.Eval script with invalid PackageName should fail immediately [2s 867ms]
  X FSharp.Compiler.Scripting.DependencyManager.UnitTests.DependencyManagerInteractiveTests.SmokeTest - #r nuget package not found [2s 843ms]
```

**Comment by dsyme** [comment]
OK, the tests are adjusted for the diagnostics updates.

File: tests/FSharp.Compiler.Private.Scripting.UnitTests/DependencyManagerInteractiveTests.fs [hunk 2]
```diff
@@ -232,7 +230,7 @@ type DependencyManagerInteractiveTests() =
 
 /// Native dll resolution is not implemented on desktop
 #if NETCOREAPP
-    [<Fact>]
+    [<Fact(Skip="downloads very large ephemeral packages"); >]
```
Comment (cartermp) [inline tests/FSharp.Compiler.Private.Scripting.UnitTests/DependencyManagerInteractiveTests.fs:233]: Should this test even exist right now? It feels like the packages there are undergoing so much churn that by the time we ever consider unskipping everything will be different

Comment (dsyme) [reply tests/FSharp.Compiler.Private.Scripting.UnitTests/DependencyManagerInteractiveTests.fs:233]: I guess we could pin the packages and they are on nuget.org forever

**Comment by dsyme** [comment]
OK, green and ready

**Review by KevinRansom — APPROVED** [review]

**Review by cartermp — APPROVED** [review]
This looks and feels good enough in VS as well.

**Pull request merged**